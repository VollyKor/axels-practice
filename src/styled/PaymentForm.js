import styled from 'styled-components';
import { RiCheckboxCircleFill } from 'react-icons/ri';

export const CheckIcon = styled(RiCheckboxCircleFill)`
    position: absolute;
    bottom: -2px;
    right: 0px;

    width: 12px;
    height: 12px;
    color: var(--white-color);
    background-color: white;
    border-radius: 50%;
`;
