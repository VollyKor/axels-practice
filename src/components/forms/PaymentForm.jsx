import { useRef } from 'react';
import { useHistory } from 'react-router-dom';
import { Form, Col, Row } from 'react-bootstrap';
import { useFormik } from 'formik';
import Autocomplete from 'react-google-autocomplete';

import { Tooltip } from 'components';
import { SubmitButton } from '../../styled/ShippingForm';
import { CheckIcon } from '../../styled/PaymentForm';
import { BiLockAlt } from 'react-icons/bi';

import {
    initialPaymentFormValues,
    fieldName,
    localStorageKey,
} from 'helpers/constants';
import { PaymentFormValidate } from '../../helpers/validationSchemas';

export default function PaymentForm() {
    const history = useHistory();

    const cardHolderRef = useRef(null);
    const cardNumberRef = useRef(null);
    const expireDateRef = useRef(null);
    const securityCodeRef = useRef(null);

    const { handleSubmit, getFieldProps, touched, errors } = useFormik({
        initialValues: initialPaymentFormValues,

        onSubmit: async (formData) => {
            localStorage.setItem(
                localStorageKey.paymentForm,
                JSON.stringify(formData)
            );

            history.push('/cart/success');
        },
        validationSchema: PaymentFormValidate,
    });

    const getWarningStyleBg = (field) =>
        errors[field] && touched[field] && 'bg-warning';

    return (
        <>
            <Row>
                <h2 className="ml-4">Payment</h2>
                <Form className="px-4 pt-1 pb-5" onSubmit={handleSubmit}>
                    <div className="text-muted d-flex align-items-baseline ">
                        <div className="position-relative mb-3">
                            <BiLockAlt size="25px" />
                            <CheckIcon fill="green" />
                        </div>
                        <small className="ml-2">
                            This is a secure 128-bit SSL encrypted payment
                        </small>
                    </div>

                    <Form.Group>
                        <Form.Label>Cardholder Name</Form.Label>
                        <Form.Control
                            ref={cardHolderRef}
                            type="name"
                            placeholder="Name as it appears on your card"
                            {...getFieldProps(fieldName.cardholder)}
                            className={getWarningStyleBg(fieldName.cardholder)}
                        />

                        <Tooltip
                            fieldName={fieldName.cardholder}
                            forwardRef={cardHolderRef}
                            touched={touched}
                            errors={errors}
                        />
                    </Form.Group>

                    <Form.Group>
                        <Form.Label>Card Number</Form.Label>
                        <Form.Control
                            ref={cardNumberRef}
                            type="tel"
                            inputMode="numeric"
                            placeholder="XXXX XXXX XXXX XXXX"
                            {...getFieldProps(fieldName.cardNumber)}
                            className={getWarningStyleBg(fieldName.cardNumber)}
                        />

                        <Tooltip
                            fieldName={fieldName.cardNumber}
                            forwardRef={cardNumberRef}
                            touched={touched}
                            errors={errors}
                        />
                    </Form.Group>

                    <Form.Row>
                        <Col md={4}>
                            <Form.Group>
                                <Form.Label>Expire Date</Form.Label>
                                <Form.Control
                                    ref={expireDateRef}
                                    type="text"
                                    placeholder="MM / YY"
                                    {...getFieldProps(fieldName.expireDate)}
                                    className={getWarningStyleBg(
                                        fieldName.expireDate
                                    )}
                                />

                                <Tooltip
                                    fieldName={fieldName.expireDate}
                                    forwardRef={expireDateRef}
                                    touched={touched}
                                    errors={errors}
                                />
                            </Form.Group>
                        </Col>

                        <Col md={4}>
                            <Form.Group>
                                <Form.Label>Security Code</Form.Label>
                                <Form.Control
                                    ref={securityCodeRef}
                                    type="text"
                                    {...getFieldProps(fieldName.securityCode)}
                                    className={getWarningStyleBg(
                                        fieldName.securityCode
                                    )}
                                />

                                <Tooltip
                                    fieldName={fieldName.securityCode}
                                    forwardRef={securityCodeRef}
                                    touched={touched}
                                    errors={errors}
                                />
                            </Form.Group>
                        </Col>
                    </Form.Row>
                    <SubmitButton type="submit" className="px-4 px-sm-5 py-2">
                        Pay Securely
                    </SubmitButton>
                </Form>
            </Row>
        </>
    );
}
