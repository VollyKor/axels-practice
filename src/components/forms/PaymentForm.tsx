import { useRef } from 'react';
import { useHistory } from 'react-router-dom';
import { Form, Col } from 'react-bootstrap';
import { useFormik } from 'formik';
import { BiLockAlt } from 'react-icons/bi';
import InputMask from 'react-input-mask';

import { Tooltip, CreditCardIcon } from 'components';
import { SubmitButton } from 'styled/ShippingForm';
import { CheckIcon } from 'styled/PaymentForm';

import { initialPaymentFormValues, fieldName } from 'helpers/constants';
import { paymentFormSchema } from 'helpers/validationSchemas';
import { localStorageKey } from 'types/enums';

export default function PaymentForm(): JSX.Element {
    const history = useHistory();

    const cardHolderRef = useRef(null);
    const cardNumberRef = useRef(null);
    const expireDateRef = useRef(null);
    const securityCodeRef = useRef(null);

    const { handleSubmit, getFieldProps, touched, errors, values } = useFormik({
        initialValues: initialPaymentFormValues,

        onSubmit: async (formData) => {
            localStorage.setItem(
                localStorageKey.paymentForm,
                JSON.stringify(formData)
            );

            history.push('/cart/success');
        },
        validationSchema: paymentFormSchema,
    });

    const getWarning: TgetWarningPaymentForm = (field) =>
        errors[field] && touched[field] && 'bg-warning';

    return (
        <>
            <h2 className="ml-4 h4">Payment</h2>
            <Form className="px-1 px-md-4 pt-1 pb-5" onSubmit={handleSubmit}>
                <div className="text-muted d-flex align-items-baseline mb-3">
                    <div className="position-relative ">
                        <BiLockAlt size="25px" />
                        <CheckIcon fill="green" />
                    </div>
                    <small className="ml-2">
                        This is a secure 128-bit SSL encrypted payment
                    </small>
                </div>

                <Form.Group>
                    <Form.Label>Cardholder Name</Form.Label>
                    <Form.Control
                        ref={cardHolderRef}
                        type="name"
                        placeholder="Name as it appears on your card"
                        {...getFieldProps(fieldName.cardholder)}
                        className={getWarning(
                            fieldName.cardholder as PaymentFormKeys
                        )}
                    />

                    <Tooltip
                        fieldName={fieldName.cardholder as PaymentFormKeys}
                        forwardRef={cardHolderRef}
                        touched={touched}
                        errors={errors}
                    />
                </Form.Group>

                <Form.Group>
                    <Form.Label>Card Number</Form.Label>
                    <div className="div" style={{ position: 'relative' }}>
                        <InputMask
                            mask="9999 9999 9999 9999"
                            {...getFieldProps(fieldName.cardNumber)}
                        >
                            {(props: any) => (
                                <Form.Control
                                    type="tel"
                                    ref={cardNumberRef}
                                    placeholder="XXXX XXXX XXXX XXXX"
                                    className={getWarning(
                                        fieldName.cardNumber as PaymentFormKeys
                                    )}
                                    {...props}
                                />
                            )}
                        </InputMask>

                        <CreditCardIcon value={values.cardNumber} />
                    </div>
                    <Tooltip
                        fieldName={fieldName.cardNumber as PaymentFormKeys}
                        forwardRef={cardNumberRef}
                        touched={touched}
                        errors={errors}
                    />
                </Form.Group>

                <Form.Row>
                    <Col md={5}>
                        <Form.Group>
                            <Form.Label>Expire Date</Form.Label>
                            <Form.Control
                                ref={expireDateRef}
                                type="text"
                                placeholder="MM / YY"
                                {...getFieldProps(fieldName.expireDate)}
                                className={getWarning(
                                    fieldName.expireDate as PaymentFormKeys
                                )}
                            />

                            <Tooltip
                                fieldName={
                                    fieldName.expireDate as PaymentFormKeys
                                }
                                forwardRef={expireDateRef}
                                touched={touched}
                                errors={errors}
                            />
                        </Form.Group>
                    </Col>

                    <Col md={5}>
                        <Form.Group>
                            <Form.Label>Security Code</Form.Label>
                            <Form.Control
                                ref={securityCodeRef}
                                type="text"
                                {...getFieldProps(fieldName.securityCode)}
                                className={getWarning(
                                    fieldName.securityCode as PaymentFormKeys
                                )}
                            />

                            <Tooltip
                                fieldName={
                                    fieldName.securityCode as PaymentFormKeys
                                }
                                forwardRef={securityCodeRef}
                                touched={touched}
                                errors={errors}
                            />
                        </Form.Group>
                    </Col>
                </Form.Row>
                <SubmitButton type="submit" className="px-4 px-sm-5 py-2">
                    Pay Securely
                </SubmitButton>
            </Form>
        </>
    );
}
