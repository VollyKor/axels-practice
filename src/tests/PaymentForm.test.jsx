import { shallow } from 'enzyme';

import PaymentForm from 'components/forms/PaymentForm';

it('renders correctly', () => {
    const wrapper = shallow(<PaymentForm />);
    expect(wrapper).toMatchSnapshot();
});
