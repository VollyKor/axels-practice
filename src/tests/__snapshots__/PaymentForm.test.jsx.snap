// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Fragment>
  <h2
    className="ml-4 h4"
  >
    Payment
  </h2>
  <Form
    className="px-1 px-md-4 pt-1 pb-5"
    inline={false}
    onSubmit={[Function]}
  >
    <div
      className="text-muted d-flex align-items-baseline mb-3"
    >
      <div
        className="position-relative "
      >
        <BiLockAlt
          size="25px"
        />
        <Styled(RiCheckboxCircleFill)
          fill="green"
        />
      </div>
      <small
        className="ml-2"
      >
        This is a secure 128-bit SSL encrypted payment
      </small>
    </div>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Cardholder Name
      </FormLabel>
      <FormControl
        name="cardholder"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="Name as it appears on your card"
        type="name"
        value=""
      />
      <Tooltip
        errors={Object {}}
        fieldName="cardholder"
        forwardRef={
          Object {
            "current": null,
          }
        }
        touched={Object {}}
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Card Number
      </FormLabel>
      <FormControl
        inputMode="numeric"
        name="cardNumber"
        onBlur={[Function]}
        onChange={[Function]}
        placeholder="XXXX XXXX XXXX XXXX"
        type="tel"
        value=""
      />
      <Tooltip
        errors={Object {}}
        fieldName="cardNumber"
        forwardRef={
          Object {
            "current": null,
          }
        }
        touched={Object {}}
      />
    </FormGroup>
    <FormRow>
      <Col
        md={5}
      >
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Expire Date
          </FormLabel>
          <FormControl
            name="expireDate"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="MM / YY"
            type="text"
            value=""
          />
          <Tooltip
            errors={Object {}}
            fieldName="expireDate"
            forwardRef={
              Object {
                "current": null,
              }
            }
            touched={Object {}}
          />
        </FormGroup>
      </Col>
      <Col
        md={5}
      >
        <FormGroup>
          <FormLabel
            column={false}
            srOnly={false}
          >
            Security Code
          </FormLabel>
          <FormControl
            name="securityCode"
            onBlur={[Function]}
            onChange={[Function]}
            type="text"
            value=""
          />
          <Tooltip
            errors={Object {}}
            fieldName="securityCode"
            forwardRef={
              Object {
                "current": null,
              }
            }
            touched={Object {}}
          />
        </FormGroup>
      </Col>
    </FormRow>
    <Styled(Button)
      className="px-4 px-sm-5 py-2"
      type="submit"
    >
      Pay Securely
    </Styled(Button)>
  </Form>
</Fragment>
`;
